jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1297 @@
+         */
+        @Override
+        public Set<E> incomingEdgesOf(V vertex)
+        {
+            return copySet(AsSynchronizedGraph.super.incomingEdgesOf(vertex));
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public Set<E> outgoingEdgesOf(V vertex)
+        {
+            return copySet(AsSynchronizedGraph.super.outgoingEdgesOf(vertex));
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean removeEdge(E e)
+        {
+            return AsSynchronizedGraph.super.removeEdge(e);
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public E removeEdge(V sourceVertex, V targetVertex)
+        {
+            return AsSynchronizedGraph.super.removeEdge(sourceVertex, targetVertex);
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean removeVertex(V v)
+        {
+            return AsSynchronizedGraph.super.removeVertex(v);
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean isCacheEnabled()
+        {
+            return false;
+        }
+    }
+
+    /**
+     * Use cache for AsSynchronizedGraph's <code>edgesOf</code>, <code>incomingEdgesOf</code>
+     * and <code>outgoingEdgesOf</code> methods.
+     */
+    private class CacheAccess
+        implements CacheStrategy<V, E>, Serializable
+    {
+        private static final long serialVersionUID = -18262921841829294L;
+
+        // A map caching for incomingEdges operation.
+        private final transient Map<V, Set<E>> incomingEdgesMap = new HashMap<>();
+
+        // A map caching for outgoingEdges operation.
+        private final transient Map<V, Set<E>> outgoingEdgesMap = new HashMap<>();
+
+        // A map caching for edgesOf operation.
+        private final transient Map<V, Set<E>> edgesOfMap = new HashMap<>();
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public E addEdge(V sourceVertex, V targetVertex)
+        {
+            E e = AsSynchronizedGraph.super.addEdge(sourceVertex, targetVertex);
+            if (e != null)
+                edgeModified(sourceVertex, targetVertex);
+            return e;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean addEdge(V sourceVertex, V targetVertex, E e)
+        {
+            if (AsSynchronizedGraph.super.addEdge(sourceVertex, targetVertex, e)) {
+                edgeModified(sourceVertex, targetVertex);
+                return true;
+            }
+            return false;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public Set<E> edgesOf(V vertex)
+        {
+            Set<E> s = edgesOfMap.get(vertex);
+            if (s != null)
+                return s;
+            synchronized (edgesOfMap) {
+                s = edgesOfMap.get(vertex);
```

Comment:
I still think that using synchronization here is better than using ConcurrentMap. If not synchronized here, a lot of copy may be created which will bring a huge memory burden.

Reply:
You are calling edgesOfMap.get outside of the synchronized block.  That can't possibly be safe, since another thread could concurrently be calling edgesOfMap.put.

And I think "huge memory burden" is an exaggeration given that

(a) only one copy will remain in edgesOfMap; others will be garbage collected as soon as the calling algorithm stops referencing it

(b) in a large graph, concurrent calls to this method are likely to be for distinct vertices